casper/src/main/scala/coop/rchain/casper/Validate.scala

```diff
@@ -141,6 +151,49 @@ object Validate {
       }
     }
 
+  /**
+    * Validate no deploy by the same (user, millisecond timestamp)
+    * has been produced in the chain
+    */
+  def repeatDeploy[F[_]: Monad: Log: BlockStore](block: BlockMessage,
+                                                 genesis: BlockMessage,
+                                                 dag: BlockDag): F[Boolean] =
+    BlockStore[F].asMap().flatMap { internalMap: Map[BlockHash, BlockMessage] =>
+      {
+        def parents(b: BlockMessage): Iterator[BlockMessage] =
+          ProtoUtil.parents(b).iterator.flatMap(internalMap.get)
+
+        val deployKeySet = block.getBody.newCode.map(d => (d.getRaw.user, d.getRaw.timestamp)).toSet
+        val iterator = DagOperations.bfTraverse(Some(genesis))(b =>
+          childMapIterator(dag.childMap.getOrElse(b.blockHash, Set.empty), internalMap))
+        val repeatedBlocks = iterator.filter(it => {
+          it.body.exists(
+            _.newCode.exists(
+              _.raw.exists(p => deployKeySet.contains((p.user, p.timestamp)))
+            )
+          )
+        })
+        // it is allowed for there to be the same (user, timestamp) transaction in both forks
+        val blockAncestors = new mutable.HashSet[BlockMessage]
+        bfTraverse[BlockMessage](Some(block))(parents).foreach(blockAncestors += _)
+
+        val invalid = repeatedBlocks.exists(b1 => {
+          blockAncestors.contains(b1)
+        })
+
+        if (!invalid) {
+          true.pure[F]
+        } else {
+          for {
+            _ <- Log[F].warn(
+                  ignore(
+                    block,
+                    "found deploy by the same (user, millisecond timestamp) produced in the chain"))
```

Comment:
It would be nice if the log had more information. Like this particular deploy was already included in this particular block.

Reply:
Done